src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/PartitionedRateLimiter.cs

```diff
@@ -182,48 +213,92 @@ private void ThrowIfDisposed()
             }
         }
 
-        private static void Replenish(DefaultPartitionedRateLimiter<TResource, TKey> limiter)
+        private async Task TimerLoop()
         {
-            lock (limiter.Lock)
+            lock (Lock)
             {
-                if (limiter._disposed)
+                if (_disposed)
                 {
                     return;
                 }
 
                 // If the cache has been invalidated we need to recreate it
-                if (limiter._cacheInvalid)
+                if (_cacheInvalid)
                 {
-                    limiter._cachedLimiters.Clear();
-                    bool cacheStillInvalid = false;
-                    foreach (KeyValuePair<TKey, Lazy<RateLimiter>> kvp in limiter._limiters)
+                    _cachedLimiters.Clear();
+                    // let's try to lower the memory held onto by the list
+                    // for example if there were 1000 items at one point, but only 10 now we don't need to keep a 1000 length buffer
+                    if (_cachedLimiters.Capacity > _limiters.Count * 2)
                     {
-                        if (kvp.Value.IsValueCreated)
-                        {
-                            if (kvp.Value.Value is ReplenishingRateLimiter)
-                            {
-                                limiter._cachedLimiters.Add(kvp.Value);
-                            }
-                        }
-                        else
+                        _cachedLimiters.Capacity = _limiters.Count;
+                    }
```

Comment:
This seems like it would thrash a lot. Wouldn't this reallocate and double the capacity if one more limiter is added? And then reallocate again to reduce size to half as soon as one more limiter is removed? If we think this is important, we should probably change this to:

```csharp
if (_cachedLimiters.Capacity > _limiters.Count * 4)
{
    _cachedLimiters.Capacity = _limiters.Count * 2;
}
```

GitHub won't let me make a suggestion because of the deleted line.

I don't think this is super necessary though. Kestrel's memory pool still doesn't shrink. It's not clear exactly when it would make sense to shrink capacity. It might seem to make sense to shrink when you have over 4 times the capacity you're currently using, but what if the traffic to the rate limiter is really bursty?

Reply:
> This seems like it would thrash a lot.

Doh, at small values it's not great. We could also limit how small we start considering lowering the size, e.g. once we're at 10 items start checking if the capacity is 2 times bigger.